File: bindings/swift/Sources/Kungfu/Kungfu.swift
// Kungfu.swift — Swift binding for the Kungfu.js framework.
//
// Uses Swift's C interop to call the C ABI directly. The native library is
// `libkungfu_core.so` / `.dylib` / `.dll`, built with
// `cargo build -p kungfu-core --features ffi`.

import Foundation
#if canImport(Glibc)
import Glibc
#elseif canImport(Darwin)
import Darwin
#endif

// ─── C ABI imports ────────────────────────────────────────────────────────────

// Opaque pointer types matching `core/kungfu.h`.
public typealias KungfuRouterRef = OpaquePointer
public typealias KungfuServerRef = OpaquePointer
public typealias KungfuRequestRef = OpaquePointer
public typealias KungfuResponseRef = OpaquePointer

public typealias KungfuHandlerFn = @convention(c) (KungfuRequestRef?, KungfuResponseRef?) -> Void

@_silgen_name("kungfu_router_new")
public func kungfu_router_new() -> KungfuRouterRef?

@_silgen_name("kungfu_router_free")
public func kungfu_router_free(_ router: KungfuRouterRef?)

@_silgen_name("kungfu_router_get")
public func kungfu_router_get(_ router: KungfuRouterRef?, _ path: UnsafePointer<CChar>, _ handler: @convention(c) (KungfuRequestRef?, KungfuResponseRef?) -> Void)

@_silgen_name("kungfu_router_post")
public func kungfu_router_post(_ router: KungfuRouterRef?, _ path: UnsafePointer<CChar>, _ handler: @convention(c) (KungfuRequestRef?, KungfuResponseRef?) -> Void)

@_silgen_name("kungfu_server_new")
public func kungfu_server_new(_ router: KungfuRouterRef?) -> KungfuServerRef?

@_silgen_name("kungfu_server_listen")
public func kungfu_server_listen(_ server: KungfuServerRef?, _ port: Int32)

@_silgen_name("kungfu_request_param")
public func kungfu_request_param(_ req: KungfuRequestRef?, _ key: UnsafePointer<CChar>) -> UnsafePointer<CChar>?

@_silgen_name("kungfu_response_status")
public func kungfu_response_status(_ res: KungfuResponseRef?, _ code: Int32)

@_silgen_name("kungfu_response_json")
public func kungfu_response_json(_ res: KungfuResponseRef?, _ json: UnsafePointer<CChar>)

// ─── Swift API ────────────────────────────────────────────────────────────────

/// A Kungfu application.
public final class Kungfu {
    private var router: KungfuRouterRef?

    public init() {
        self.router = kungfu_router_new()
    }

    deinit {
        if let r = router { kungfu_router_free(r) }
    }

    public func get(_ path: String, handler: @escaping (Request, Response) -> Void) {
        path.withCString { cPath in
            // V1 scaffold: the C callback can't capture Swift closures
            // directly. We use a global handler table keyed by path.
            HandlerTable.shared.register(path, handler: handler)
            kungfu_router_get(router, cPath) { req, res in
                let request = Request(ptr: req!)
                let response = Response(ptr: res!)
                // Look up the handler by reading the request path.
                // For V1 simplicity, just call the first registered handler.
                if let h = HandlerTable.shared.handlers.first?.value {
                    h(request, response)
                }
            }
        }
    }

    public func listen(_ port: Int = 3000) {
        let server = kungfu_server_new(router)
        kungfu_server_listen(server, Int32(port))
    }
}

public struct Request {
    let ptr: KungfuRequestRef

    public func param(_ key: String) -> String {
        return key.withCString { cKey -> String in
            guard let cVal = kungfu_request_param(ptr, cKey) else { return "" }
            return String(cString: cVal)
        }
    }
}

public struct Response {
    let ptr: KungfuResponseRef

    public func status(_ code: Int) -> Response {
        kungfu_response_status(ptr, Int32(code))
        return self
    }

    public func json(_ json: String) {
        json.withCString { cJson in
            kungfu_response_json(ptr, cJson)
        }
    }
}

// V1: global handler table (workaround for C-callback-can't-capture-closure).
internal class HandlerTable {
    static let shared = HandlerTable()
    var handlers: [String: (Request, Response) -> Void] = [:]
    func register(_ path: String, handler: @escaping (Request, Response) -> Void) {
        handlers[path] = handler
    }
}

File: bindings/dart/lib/kungfu.dart
// kungfu.dart — Dart binding for the Kungfu.js framework.
//
// Uses dart:ffi to call the C ABI. The native library is `libkungfu_core.so`
// (Linux), `libkungfu_core.dylib` (macOS), or `kungfu_core.dll` (Windows),
// built with `cargo build -p kungfu-core --features ffi`.

import 'dart:ffi';
import 'dart:io';
import 'package:ffi/ffi.dart';

// ─── Native bindings ──────────────────────────────────────────────────────────

typedef _NativeRouterNew = Pointer<Void> Function();
typedef _DartRouterNew = Pointer<Void> Function();

typedef _NativeRouterFree = Void Function(Pointer<Void>);
typedef _DartRouterFree = void Function(Pointer<Void>);

typedef _NativeRouterGet = Void Function(Pointer<Void>, Pointer<Utf8>, Pointer<NativeFunction<_NativeHandler>>);
typedef _DartRouterGet = void Function(Pointer<Void>, Pointer<Utf8>, Pointer<NativeFunction<_NativeHandler>>);

typedef _NativeHandler = Void Function(Pointer<NativeRequest>, Pointer<NativeResponse>);
typedef _DartHandler = void Function(Pointer<NativeRequest>, Pointer<NativeResponse>);

typedef _NativeServerListen = Void Function(Pointer<Void>, Int32);
typedef _DartServerListen = void Function(Pointer<Void>, int);

typedef _NativeRequestParam = Pointer<Utf8> Function(Pointer<NativeRequest>, Pointer<Utf8>);
typedef _DartRequestParam = Pointer<Utf8> Function(Pointer<NativeRequest>, Pointer<Utf8>);

typedef _NativeResponseStatus = Void Function(Pointer<NativeResponse>, Int32);
typedef _DartResponseStatus = void Function(Pointer<NativeResponse>, int);

typedef _NativeResponseJson = Void Function(Pointer<NativeResponse>, Pointer<Utf8>);
typedef _DartResponseJson = void Function(Pointer<NativeResponse>, Pointer<Utf8>);

// Opaque structs — we only need the pointer types.
class NativeRequest extends Opaque {}
class NativeResponse extends Opaque {}

/// Loads the native Kungfu library.
final DynamicLibrary _lib = _loadLib();

DynamicLibrary _loadLib() {
  if (Platform.isLinux) return DynamicLibrary.open('libkungfu_core.so');
  if (Platform.isMacOS) return DynamicLibrary.open('libkungfu_core.dylib');
  if (Platform.isWindows) return DynamicLibrary.open('kungfu_core.dll');
  throw UnsupportedError('Unsupported platform');
}

// ─── Dart API ─────────────────────────────────────────────────────────────────

class Kungfu {
  late final Pointer<Void> _router;

  Kungfu() {
    final newFn = _lib.lookupFunction<_NativeRouterNew, _DartRouterNew>('kungfu_router_new');
    _router = newFn();
  }

  void get(String path, void Function(Request, Response) handler) =>
      _register('kungfu_router_get', path, handler);

  void post(String path, void Function(Request, Response) handler) =>
      _register('kungfu_router_post', path, handler);

  void put(String path, void Function(Request, Response) handler) =>
      _register('kungfu_router_put', path, handler);

  void delete(String path, void Function(Request, Response) handler) =>
      _register('kungfu_router_delete', path, handler);

  void listen(int port) {
    final listenFn = _lib.lookupFunction<_NativeServerListen, _DartServerListen>('kungfu_server_listen');
    listenFn(_router, port);
  }

  void _register(String symbol, String path, void Function(Request, Response) handler) {
    final fn = _lib.lookupFunction<_NativeRouterGet, _DartRouterGet>(symbol);
    final pathPtr = path.toNativeUtf8();
    // V1 limitation: we don't yet wire per-route handlers through dart:ffi.
    // For now, this is a scaffold. The lookup of native callbacks requires
    // `Pointer.fromFunction`, which needs a top-level function — not a
    // closure. This is a known Dart FFI limitation we'll work around in V1.1.
    // fn(_router, pathPtr, ...);
    calloc.free(pathPtr);
    // Stash the handler so it doesn't get GC'd.
    _handlers[path] = handler;
  }

  // Stash handlers (V1 scaffold).
  static final Map<String, void Function(Request, Response)> _handlers = {};
}

class Request {
  final Pointer<NativeRequest> _ptr;
  Request(this._ptr);

  String param(String key) {
    final fn = _lib.lookupFunction<_NativeRequestParam, _DartRequestParam>('kungfu_request_param');
    final keyPtr = key.toNativeUtf8();
    final resultPtr = fn(_ptr, keyPtr);
    calloc.free(keyPtr);
    final value = resultPtr.address == 0 ? '' : resultPtr.toDartString();
    return value;
  }
}

class Response {
  final Pointer<NativeResponse> _ptr;
  Response(this._ptr);

  Response status(int code) {
    final fn = _lib.lookupFunction<_NativeResponseStatus, _DartResponseStatus>('kungfu_response_status');
    fn(_ptr, code);
    return this;
  }

  void json(String jsonStr) {
    final fn = _lib.lookupFunction<_NativeResponseJson, _DartResponseJson>('kungfu_response_json');
    final ptr = jsonStr.toNativeUtf8();
    fn(_ptr, ptr);
    calloc.free(ptr);
  }
}
